### ISSUE TITLE
explicit annotation for `defaultdict(float)` is not accepted

### ISSUE THREAD

**Issue by dimbleby** [astral-sh/ty#2196]
### Summary

Some sort of `float | int` vs `float` confusion - I see the related FAQ - I did not find any duplicate issues but perhaps there is one out there

```python
from collections import defaultdict

foo: dict[str, float] = defaultdict(float)
```
with result
```
error[invalid-assignment]: Object of type `defaultdict[Unknown, float]` is not assignable to `dict[str, int | float]`
 --> foo.py:3:6
  |
1 | from collections import defaultdict
2 |
3 | foo: dict[str, float] = defaultdict(float)
  |      ----------------   ^^^^^^^^^^^^^^^^^^ Incompatible value of type `defaultdict[Unknown, float]`
  |      |
  |      Declared type
  |
info: rule `invalid-assignment` is enabled by default

Found 1 diagnostic
```

awkward if I am also trying to annotate my code in a way that will satisfy mypy, which often insists that it needs a type annotation in such cases

### Version

0.0.6

**Comment by carljm** [comment]
Thanks for the report! Yeah, we'll either need to relax the rules around invariance for `float` vs `float | int`, or make sure we always have type context available in such cases.

I think this alternative might make both mypy and ty happy for now?

```py
foo = defaultdict[str, float](float)

**Comment by AlexWaygood** [comment]
Does this still repro on the branch for https://github.com/astral-sh/ruff/pull/21930?

**Comment by carljm** [comment]
> Does this still repro on the branch for [astral-sh/ruff#21930](https://github.com/astral-sh/ruff/pull/21930)?

Good call! The last few issues I tried, that PR didn't fix, but it makes sense that it would fix this one, and it does!

**Comment by carljm** [comment]
Closing as dupe of #1576, since the PR to close that issue fixes this.
